### ISSUE TITLE
Patch provided to make multiple process of bear collaborate

### ISSUE THREAD

**Issue by asierllano** [rizsotto/Bear#509]
When you run multiple bear instances with the --append flag
and the same output file someone could expect all the processes data
to be included in the same output seamlessly.

Additionally, if one of the process is terminated before finishing,
someone could expect that the output is either updated or not updated,
but the result is not corrupted with partially written data.

Current implementation of bear doesn't work this way, multiple processes
collide and an abrupt termination will provide a corrupted output.

I provide a patch that fixes both.

Here are the details of the change:
 - The events.json file includes the pid of the process so different bear
   instances dont' overlap one to the other, using their own event file.
 - The output object is modified to have locking mechanism and be atomic:
    - The path of the output file is passed in the constructor in order
      to make it part of the object for simplification.
    - Before starting to read a file that we will append, we get a lock
      to ensure that we are the only one appending data in that file.
    - We check the existance of the file after the lock is obtained, for
      the condition in which we are the first process to append.
    - The output is created in a temporary object and replace it atomically
      with a rename operation to avoid having the possibility that a bear
      instance terminated in the middle creates a corrupted output.
 - The application object doesn't check the existance of the output file
   when called with append flag, as it will be done later in a more thread safe
   way.
 - Created a Lock object that encapsulates in C++ the locking mechanism.

I am testing it in a complex closed source project, with many bear processes in a heterogeneous way. We don't want to run bear against everything that we do to avoid the output json file to be huge, but it is run several times. It is hard for us to separate all our project in the one bellow bear and everything else. For us it is important that the output of the json file of all the processes end up in the same file in order for the visual studio code to apply its intellisense logic properly to all the code. With the provided patch it worked for many minutes of compilation with 12 simultaneous processes, stopping them randomly and continuing without any crash. Before this patch, we had 100% of probablity of ending up with non-stopping solution that wasted 100% of the CPU or with corrupted outputs. Only worked with just 1 single process.

PATCH provided:
[Bear-3.0.21-multiprocess.patch](https://github.com/rizsotto/Bear/files/10707182/Bear-3.0.21-multiprocess.patch)

**Comment by rizsotto** [comment]
Hey @asierllano , thanks for your message.

I am interested in your use case. Could you explain why the different runs of Bear? (Are they running different compilers? Are they different with the compiler flags?) What do you use the output file for? (`clang-tidy` or what is the tool that you run against the compilation database?) Do you use this in CI or in your development environment?

From your explanation it seems that you want to filter out files that not your interest. Are you familiar with the [configuration file](https://github.com/rizsotto/Bear/blob/master/source/citnames/man/bear-citnames.1.md#config-file), where you can specify which files you want to include or exclude from the output?

And could your use case be covered to run the `bear intercept` and `bear citnames` programs separately? (If you are not familiar with those, [this page](https://github.com/rizsotto/Bear/wiki/Troubleshooting#how-it-works) explains the details.) Is it possible to run the `bear intercept` for your targets, concatenate the event files and run a single `bear citnames` at the end?

**Comment by asierllano** [comment]
*Explanation of my use case*
We have a custom development environment for embedded development that
performs crosscompilation. The system itself uses many processes and with a
complex Makefile is able to build many things. Some of these things are
part of the code that are integrated into the embedde device, but many of
them are tools. These tools include the crosscompiler itself. The json
generated by bear is exported so that the code can use the Intellisense
functionality of Visual Studio Code that is able to navigate properly
through the code and  highlight the compiling code (based on the enabled
MACROs). The main Makefile is a package based system. It can invoke the
build processes of autotools, cmake, and other internal tools. But it does
now if the target package is a tool or code to crosscompile.

*Why not calling a single instance of bear*
The main problem is the volume of source code, which makes things very
slow. The generated json files are around 300MB.
But more than 90% of the code is not interesting for this functionality. We
are only interested in intellisense for the code that is being
crosscompiled. We are not interested in having Intellisense in the
crosscompiler itself and other huger tools that are not the place where
most effort is required.

*Next steps:*
The provided patch worked really well!!
But if you want I can try different options.
I would discard using the configuration files, because that would mean that
I would be taking events for most part of the building process to discard
them, wasting 300MB of events and making things slower than required.
But I think that calling intercept everywhere I am calling bear now,
concatenate them and calling citnames afterwards will be a good solution,
more efficient than it is now.

I will try and let you know.

Best regards,
Asier

**Comment by asierllano** [comment]
You were right!

*Why the method you proposed is better:*
Just invoking the "intercept" myself, concatenating the needed files and
calling "citnames" improved a little bit my previous solution.
The scripting with the multi-process-safe solution was much simpler, but
the call-it-by-hand solution is more powerful.
Reason why it is more complicated:
 - Everywhere I called bear, I have to intercept the output and then I have
to put a proper output for the interception.
 - I have created a new step that calles citnames and it has to know the
names of the outputs intercepted. In my process it was not evident where to
generate this output and how to make sure that it is properly generated
even if the compilation prcess fails (because the Makefile stops).
Reasons why it is now more powerful:
 - I had several binaries and now I can differentiate one with another. It
is also supported by the VSCode IDE. So now the developer can select which
binary to intellisense and it works properly even if some of the files are
compiled with different flags.
Reason why it is faster:
 - Now I don't generate the json output all the time. So it makes the
process a bit faster.

*Still a couple a ways to improve the tool:*
- I have to generate a tempoerary file that is the concatenation of all the
inputs. Maybe we could improve "citnames" in order to accept '-' as the
input. This way I would save around 100MB of hard drive for each developer,
which is useless (just to concatenate other existing files).
- If someone stops the compilation during the call at citnames, the output
is still corrupt. This would require at least part of the patch I provided
to you.

*Questions:*
Are you still interested in the multi-process-safe approach or are you
going to discard it?
Do you think that a patch for citnames to accept the input '-' to use the
standard would be useful? I can provide it.
What about the output being generated to a temporary with overwrite rename
so that its contents is transactional and cannot be corrupt?

Best regards,
Asier

**Comment by rizsotto** [comment]
Hey @asierllano , thanks for coming back. I am glad that you've found a good solution to your use case.

> Are you still interested in the multi-process-safe approach or are you going to discard it?

I see the locking is an extra complexity, and if not needed I would not include it.

> Do you think that a patch for `citnames` to accept the input `-` to use the standard would be useful? I can provide it.

Yes, that's sound a reasonable improvement. Sure, pull requests are always welcome! :smile: 

> What about the output being generated to a temporary with overwrite rename so that its contents is transactional and cannot be corrupt?

For this I need to understand (or reproduce) this problem. The current implementation collects the compilation database entries in memory first, then it writes down to the disk. If the process run is interrupted when the write happens it leaves an incomplete file. In other cases the file was not touched (when run with `--append`), or not exists. My guess that you saw "corrupt" files when you interrupted the serialization itself. Is this correct?

Yes, writing into a temporary file and rename afterwards will fix this problem. Could you make it as a separate PR (from the standard output read)?

**Comment by asierllano** [comment]
Thank you very much.

It is now working really smooth. In the process I have found a couple of
bugs, related to data corruption in very stressing conditions.
Bear in mind, that I have around 80 different intercept calls, around 12 of
them simultaneously.
Even if they were called with different inputs and outputs, they were
supposed to work smoothly.
But the fact is that these bugs were preventing the system working in some
situations.
I have already uploaded each patch in a separated issue in Github. Feel
free to comment on them.

Now, for my use case, with the provided patches, it is stable as a rock.

Best regards,
Asier
